RHDEVDOCS-7665: Content creation for GitOps 1.21.0 RN (#113350)
openshift/openshift-docs: modules/gitops-release-notes-1-21-0.adoc

+
[source,terminal]
----
$ oc rollout restart deployment/argocd-image-updater-controller
----
+
* Increase pod PID limits.
+
link:https://issues.redhat.com/browse/GITOPS-9959[GITOPS-9959]

[id="breaking-changes-1-21-0_{context}"]
== Breaking changes

ApplicationSet tokenRef strict mode enabled by default::
With this update, the {gitops-title} Operator enables ApplicationSet `tokenRef` strict mode by default. This happens when `.spec.applicationSet.sourceNamespaces` on the Argo CD custom resource includes at least one cluster namespace. This change affects ApplicationSet SCM Provider and Pull Request generators. Secrets used by these generators through `tokenRef` must have this label: `argocd.argoproj.io/secret-type: scm-creds`.
+
If you use SCM Provider or Pull Request generators with unlabeled secrets, do one of the following before upgrading:
+
* Recommended: Label your secrets with `argocd.argoproj.io/secret-type: scm-creds`
+
* Not recommended for production environments: Temporarily disable strict mode by setting the following configuration in the ArgoCD custom resource:
+
[source,yaml]
----
spec:
  cmdParams:
    applicationsetcontroller:

>       enable.tokenref.strict.mode: "false"

Reviewer: ```suggestion
  cmdParams:
    applicationsetcontroller.enable.tokenref.strict.mode: "false"
```

Author: Fixed.

== Errata updates

RHEA-XXXX:NNNN - {gitops-title} 1.21.0 enhancement update advisory::
Issued: 2026-NN-XX
+
The enhancements included in this release are documented in the following advisory:
+
* link:https://access.redhat.com/errata/RHEA-XXXX:NNNN[RHEA-XXXX:NNNN]

If you installed the {gitops-title} Operator in the default namespace, run the following command to view the container images in this release:

[source,terminal]
----
$ oc describe deployment gitops-operator-controller-manager -n openshift-gitops-operator
----

[id="new-features-1-21-0_{context}"]
== New features

Argo CD Image Updater (General Availability)::
With this update, Argo CD Image Updater is generally available. It updates container images automatically for Kubernetes workloads managed by Argo CD. It tracks image versions in applications. You can define these versions in `ImageUpdater` custom resources (CR). It updates images using parameter overrides through the Argo CD API or the Git write-back method. The Image Updater supports applications built with Kustomize or Helm.
+
link:https://issues.redhat.com/browse/GITOPS-9754[GITOPS-9754]

GitOps Console Plugin resource pages (Technology Preview)::

> With this update, the GitOps Console Plugin adds support for Argo CD and Argo Rollouts as a Technology Preview. You can manage Applications, ApplicationSets, AppProjects, and Rollouts directly in the Console.

Reviewer 2: There are some things missing here.

I actually prefer the contents in the release note text from 9196.

Author: Fixed. We don't really have the Administrator perspective in the UI so I removed that part.

+
This makes GitOps operations easier.
+
--
:FeatureName: The GitOps Console Plugin resource pages feature
include::snippets/technology-preview.adoc[]
--
+
link:https://issues.redhat.com/browse/GITOPS-9196[GITOPS-9196]

Argo CD CLI (General Availability)::
With this update, the Argo CD CLI is generally available. It provides a stable command-line interface for managing Argo CD applications and related resources. Use the CLI to perform common GitOps operations.
+
link:https://issues.redhat.com/browse/GITOPS-9379[GITOPS-9379]

Multi-source Applications support in Argo CD UI::
With this update, Argo CD supports creating multi-source Applications directly from the New Application UI. Before this update, multi-source Applications (`spec.sources`) could only be created through the CLI, manifests, or manual YAML editing in the UI. The guided New Application panel supported only a single source. Users had to switch to YAML editing to combine multiple Git, Helm, or OCI sources. With this update, you can add, configure, and remove multiple sources directly from the UI.
+
link:https://issues.redhat.com/browse/GITOPS-9407[GITOPS-9407]

Pull Request workflow support for Image Updater::

> With this update, Argo CD Image Updater adds a Pull Request mode for the Git write-back method. Before this update, image update commits were pushed directly to the tracking branch. This did not work with protected branches and skipped review processes. With this update, the Image Updater can open a pull request to the base branch instead of pushing directly. This allows image updates to go through a review and approval workflow.

Reviewer 3: we can remove mode i guess.

Author: removed the word mode as per your suggestion. Fixed.

+
link:https://issues.redhat.com/browse/GITOPS-9143[GITOPS-9143]

Namespace-scoped operation for Image Updater::

> With this update, Argo CD Image Updater defaults to namespace-scoped operation. Before this update, the Image Updater required cluster-wide RBAC permissions by default. This was restrictive in multitenant environments where cluster-scoped access is limited. The controller now watches only the namespace where it is installed. It includes namespace-scoped Role and RoleBinding manifests. You can still configure the controller to watch a specific set of namespaces or the entire cluster. Adjust the RBAC configuration to do this.

Reviewer 3: IMO, it is better to mention it as default behavior as it can be configurable.

Author: Fixed.

> With this update, Argo CD Image Updater defaults to namespace-scoped operation. Before this update, the Image Updater required cluster-wide RBAC permissions by default. This was restrictive in multitenant environments where cluster-scoped access is limited. The controller now watches only the namespace where it is installed. It includes namespace-scoped Role and RoleBinding manifests. You can still configure the controller to watch a specific set of namespaces or the entire cluster. Adjust the RBAC configuration to do this.

Reviewer 3: multi-tenant

Author: I would like to keep this as multitenant.

+
link:https://issues.redhat.com/browse/GITOPS-9134[GITOPS-9134]

TLS encryption for Image Updater webhook server::

> With this update, the Argo CD Image Updater webhook server runs with TLS enabled by default. Before this update, the webhook server accepted registry notifications over plain HTTP. It had no TLS encryption or configuration options. This allowed webhook payloads to transmit in cleartext. This included shared secrets used for payload validation. The webhook server now uses TLS 1.3 for both the minimum and maximum protocol versions. This ensures secure transmission of webhook notifications from container registries.

Reviewer 3: TLS 1.3 is default values right ?

Reviewer 4: yes. We should make it clear that "The webhook server now uses TLS 1.3 for both the minimum and maximum protocol versions _by default_". And these tls options are configurable.

Author: Fixed.

+
link:https://issues.redhat.com/browse/GITOPS-8784[GITOPS-8784]

SSH handshake timeout errors with GitHub App authentication behind firewalls::
Before this update, users with GitHub App authentication behind restrictive firewalls saw vague timeout errors. This happened because Post-Quantum Cryptography (PQC) algorithms generate larger SSH handshake packets. Some firewalls silently drop these packets. PQC algorithms are enabled by default starting in Argo CD 2.4.12. Suricata 7, which AWS Network Firewall uses, drops such large SSH packets. With this update, individual timeouts are used. Error messages now explicitly identify SSH handshake timeouts instead of generic timeout errors. This helps cluster operators quickly isolate and debug network path issues.
+
To avoid timeout errors when using GitHub App authentication behind network firewalls, disable the PQC algorithm for SSH handshake.
+
link:https://issues.redhat.com/browse/GITOPS-8739[GITOPS-8739]

[id="known-issues-1-21-0_{context}"]
== Known issues

Goroutine thread leak in Image Updater with Git write-back method::

> The Argo CD Image Updater might have a thread leak when using the `git` write-back method under heavy load. This happens when you have many applications and short poll intervals. The leak causes Git operations to fail with this error: `error: cannot fork() for remote-https: Resource temporarily unavailable`.

Reviewer 3: instead of this, can we mention goroutine thread or go routine ?

Author: Fixed.

+
Use the camelCase `logFormat` field for `applicationSet` and `notifications`. The lowercase `logformat` field is deprecated but still works for backward compatibility. The {gitops-title} Operator uses the lowercase `logformat` field if the camelCase `logFormat` field is not set. You can update your custom resource specifications to use the camelCase `logFormat` field. The deprecated lowercase `logformat` field might be removed in a future release.
+
link:https://issues.redhat.com/browse/GITOPS-9119[GITOPS-9119]

Deprecated dynamic shard scaling feature::
With this update, the dynamic shard scaling feature in the {gitops-title} Operator is deprecated. This includes the `.spec.controller.sharding.dynamicScalingEnabled` field in the Argo CD custom resource. For more information about sharding clusters, see the Additional resources section.
+
link:https://issues.redhat.com/browse/GITOPS-9478[GITOPS-9478]

Removed kube-rbac-proxy dependency::

> With this update, the deprecated `kube-rbac-proxy` image is no longer used. This affects three operators: `argocd-operator`, `gitops-operator`, and `argocd-rollouts-manager`. These operators now use the `WithAuthenticationAndAuthorization` function from `controller-runtime`. Operator deployments now run with one container instead of two.

Reviewer 5: ```suggestion
With this update, the deprecated `kube-rbac-proxy` image is no longer used. The {gitops-title}  Operator now uses the `WithAuthenticationAndAuthorization` function from `controller-runtime` for authentication and authorization. As a result, operator deployments now run with a single container instead of two.
```
Lets not mention about argocd and rollouts operator

Author: Fixed.

+
link:https://issues.redhat.com/browse/GITOPS-9701[GITOPS-9701]

End-to-end TLS encryption for Redis traffic in Argo CD Agent and Principal::
With this update, the Argo CD Agent and Principal components support end-to-end TLS encryption for Redis traffic. The Redis proxy accepts TLS connections from Argo CD components like `argocd-server` and `argocd-repo-server`. It uses TLS when connecting to the local `argocd-redis` instance. Configure this feature using new CLI flags and environment variables. These include `--redis-tls-enabled`, proxy server certificate and key paths, or related Kubernetes secrets. You can configure the upstream CA using a file path, a Kubernetes secret, or an insecure mode for testing only. The Kubernetes and Helm installation manifests include optional Redis TLS settings. TLS is disabled by default. The `principal.redis.tls.enabled` and `agent.redis.tls.enabled` fields default to `false`. This ensures that existing deployments continue to work until you enable the feature and provide certificates.

> +

Reviewer 6: LGTM

link:https://issues.redhat.com/browse/GITOPS-8090[GITOPS-8090]

Image Updater status subresource for monitoring and observability::
With this update, each `ImageUpdater` CR maintains a status subresource that reflects the overall state of the resource. This includes the number of matched applications, managed images, timestamps, conditions, and a list of image updates from the most recent update cycle. You can check the state of your `ImageUpdater` resources with CLI commands:
+
[source,terminal]
----
$ oc get imageupdater -n openshift-gitops
----
+
You can also view the complete status by examining the `ImageUpdater` CR in the cluster.
+
link:https://issues.redhat.com/browse/GITOPS-7316[GITOPS-7316]

[id="fixed-issues-1-21-0_{context}"]
== Fixed issues

> Dex service account authentication security tokens::

Reviewer 6: LGTM

Before this update, the Dex service account used non-expiring security tokens. This created security risks from long-lived credentials. With this update, the service account uses expiring tokens. This improves authentication security for the Dex server in {gitops-title}.
+
link:https://issues.redhat.com/browse/GITOPS-9426[GITOPS-9426]

> Dex copyutil initContainer reconciliation during upgrades::

Reviewer 6: LGTM

+
link:https://issues.redhat.com/browse/GITOPS-8739[GITOPS-8739]

[id="known-issues-1-21-0_{context}"]
== Known issues

Goroutine thread leak in Image Updater with Git write-back method::
The Argo CD Image Updater might have a thread leak when using the `git` write-back method under heavy load. This happens when you have many applications and short poll intervals. The leak causes Git operations to fail with this error: `error: cannot fork() for remote-https: Resource temporarily unavailable`.
+
To mitigate this issue, apply one of the following workarounds:
+
* Increase the reconciliation interval to slow the leak by setting the environment variable in the ArgoCD custom resource:
+
[source,yaml]
----
spec:
  imageUpdater:
    env:
      - name: IMAGE_UPDATER_INTERVAL
        value: 15m
----
+
* Implement a periodic restart of the deployment:
+
[source,terminal]
----
$ oc rollout restart deployment/argocd-image-updater-controller
----
+
* Increase pod PID limits.
+
link:https://issues.redhat.com/browse/GITOPS-9959[GITOPS-9959]

[id="breaking-changes-1-21-0_{context}"]

> == Breaking changes

Reviewer 7: Could you add an Image Updater Breaking change:
https://argocd-image-updater.readthedocs.io/en/stable/#argo-cd-image-updater
https://argocd-image-updater.readthedocs.io/en/stable/#namespace-field-removal

```
Breaking Change: spec.namespace Field Removed

The spec.namespace field has been removed from the ImageUpdater API. Any CR that includes spec.namespace will fail validation. The controller uses the ImageUpdater CR's metadata.namespace to determine which namespace to search for applications. For details and migration examples, see [Namespace Field Removal](https://argocd-image-updater.readthedocs.io/en/stable/#namespace-field-removal) below.
```

Author: @dkarpele, sure we can add that. Could you please provide the JIRA ticket too? Thanks :)

Reviewer 7: 9135

Author: Fixed.

With this update, the Argo CD Agent and Principal components support end-to-end TLS encryption for Redis traffic. The Redis proxy accepts TLS connections from Argo CD components like `argocd-server` and `argocd-repo-server`. It uses TLS when connecting to the local `argocd-redis` instance. Configure this feature using new CLI flags and environment variables. These include `--redis-tls-enabled`, proxy server certificate and key paths, or related Kubernetes secrets. You can configure the upstream CA using a file path, a Kubernetes secret, or an insecure mode for testing only. The Kubernetes and Helm installation manifests include optional Redis TLS settings. TLS is disabled by default. The `principal.redis.tls.enabled` and `agent.redis.tls.enabled` fields default to `false`. This ensures that existing deployments continue to work until you enable the feature and provide certificates.
+
link:https://issues.redhat.com/browse/GITOPS-8090[GITOPS-8090]

Repository credential sync for Argo CD Agent::
With this update, Argo CD Agent syncs credentials to managed agents. It supports per-repository secrets and URL-prefix templates. Both are scoped by AppProject.
+

> link:https://issues.redhat.com/browse/GITOPS-9390[GITOPS-9390], link:https://issues.redhat.com/browse/GITOPS-9022[GITOPS-9022]

Reviewer 8: https://issues.redhat.com/browse/GITOPS-9022 is part of the Hybrid Architecture note below

Author: Ok, so I am only keeping it as GITOPS-9390 now. Removed 9022 from the content.

Hybrid architecture for Argo CD Agent (Technology Preview)::

> With this update, you can run Argo CD Agent with existing Argo CD. This lets you adopt the agent pull model step by step. Traditional applications use the Application Controller. Agent applications sync to workload clusters via agent pull model. Both appear in the same UI. Label selectors enforce isolation between them.

Reviewer 8: Minor changes to enhance accuracy. Feel free to update.

``` suggestion
With this update, you can run Argo CD Agent with the existing Argo CD setup. This lets you adopt the agent pull model step by step without replacing the existing push-based Argo CD setup. Traditional applications continue to use the Application Controller on the control plane. Agent applications are synced to the workload clusters via the agent pull model. Both appear in the same UI. Label selectors and the `argocd.argoproj.io/skip-reconcile` annotation enforce isolation between them.
```

Author: Fixed.

+
link:https://issues.redhat.com/browse/GITOPS-8090[GITOPS-8090]

Repository credential sync for Argo CD Agent::
With this update, Argo CD Agent syncs credentials to managed agents. It supports per-repository secrets and URL-prefix templates. Both are scoped by AppProject.
+
link:https://issues.redhat.com/browse/GITOPS-9390[GITOPS-9390]

Hybrid architecture for Argo CD Agent (Technology Preview)::
With this update, you can run Argo CD Agent with the existing Argo CD setup. This lets you adopt the agent pull model step by step without replacing the existing push-based Argo CD setup. Traditional applications continue to use the Application Controller on the control plane. Agent applications are synced to the workload clusters via the agent pull model. Both appear in the same UI. Label selectors and the `argocd.argoproj.io/skip-reconcile` annotation enforce isolation between them.
+
--
:FeatureName: The Argo CD Agent hybrid architecture feature
include::snippets/technology-preview.adoc[]
--
+

> link:https://issues.redhat.com/browse/GITOPS-9790[GITOPS-9022]

Reviewer 8: The link points to a different story
```suggestion
link:https://issues.redhat.com/browse/GITOPS-9022[GITOPS-9022]
```

Author: Sorry about this, it is now fixed.

== Errata updates

RHEA-XXXX:NNNN - {gitops-title} 1.21.0 enhancement update advisory::
Issued: 2026-NN-XX
+
The enhancements included in this release are documented in the following advisory:
+
* link:https://access.redhat.com/errata/RHEA-XXXX:NNNN[RHEA-XXXX:NNNN]

If you installed the {gitops-title} Operator in the default namespace, run the following command to view the container images in this release:

[source,terminal]
----
$ oc describe deployment gitops-operator-controller-manager -n openshift-gitops-operator
----

[id="new-features-1-21-0_{context}"]
== New features

Argo CD Image Updater (General Availability)::
With this update, Argo CD Image Updater is generally available. It updates container images automatically for Kubernetes workloads managed by Argo CD. It tracks image versions in applications. You can define these versions in `ImageUpdater` custom resources (CR). It updates images using parameter overrides through the Argo CD API or the Git write-back method. The Image Updater supports applications built with Kustomize or Helm.
+
link:https://issues.redhat.com/browse/GITOPS-9754[GITOPS-9754]

GitOps Console Plugin resource pages (Technology Preview)::

> With this update, the GitOps Console Plugin adds Technology Preview support for Argo CD and Argo Rollouts. You can manage Applications, ApplicationSets, AppProjects, and Rollouts directly from the {OCP}.

Reviewer 2: "You can manage...."

At this point, we aren't ready to allow users to perform certain key "managing" actions, like sync an application or promote a rollout.  

Instead, we allow users to view these resources and statuses quickly, especially for those users who want to continue to use the OpenShift Console.  If they want they can then click on links that we provide to the Argo CD UI where they can do more actions.

Instead, rephrase it to:

"You can view Applications, ApplicationSets...."

I updated the RN text in the Jira.  I think when we add more actions, then we can change it back to "You can manage..."

Author: Fixed.

+
This feature includes:
+
* List and details pages for Argo CD and Argo Rollouts resources
* Filters for health and sync status
* Visual views for Applications and ApplicationSets
* Rollout topology in the OpenShift Console Topology view
* YAML templates for creating resources
* Links between the OpenShift Console and the Argo CD UI
* Views for sync history, events, resources, generators, revisions, and project policies

> +

Reviewer 2: LGTM as a general description, regarding the feature list

Author: Thanks.

These updates help you manage {gitops-shortname} resources from a single interface and reduce the need to switch between tools.
+
--
:FeatureName: The GitOps Console Plugin resource pages feature
include::snippets/technology-preview.adoc[]
--
+
link:https://issues.redhat.com/browse/GITOPS-9196[GITOPS-9196]

Argo CD CLI (General Availability)::
With this update, the Argo CD CLI is generally available. It provides a stable command-line interface for managing Argo CD applications and related resources. Use the CLI to perform common GitOps operations.
+
link:https://issues.redhat.com/browse/GITOPS-9379[GITOPS-9379]

Multi-source Applications support in Argo CD UI::

> With this update, Argo CD supports creating multi-source Applications directly from the New Application UI. Before this update, multi-source Applications (`spec.sources`) could only be created through the CLI, manifests, or manual YAML editing in the UI. The guided New Application panel supported only a single source. Users had to switch to YAML editing to combine multiple Git, Helm, or OCI sources. With this update, you can add, configure, and remove multiple sources directly from the UI.

Reviewer 2: Regarding the last sentence "With this update, you can add, configure, and remove multiple sources directly from the UI."

The new work that Atif did in 9407 was to change the New App panel to support creating apps with multiple sources.  Before, if you had a multi-source app, in the Details panel's Sources Tab of the App, you can still edit, add or delete sources from the app.  This support isn't new in the UI.

So, the idea is, "With this update, you can now CREATE a new application (from scratch) with multiple sources via the New App panel in the Argo CD UI" or something like that

Author: Fixed.

+
link:https://issues.redhat.com/browse/GITOPS-8090[GITOPS-8090]

Repository credential sync for Argo CD Agent::
With this update, Argo CD Agent syncs credentials to managed agents. It supports per-repository secrets and URL-prefix templates. Both are scoped by AppProject.
+
link:https://issues.redhat.com/browse/GITOPS-9390[GITOPS-9390], link:https://issues.redhat.com/browse/GITOPS-9022[GITOPS-9022]

Hybrid architecture for Argo CD Agent (Technology Preview)::
With this update, you can run Argo CD Agent with existing Argo CD. This lets you adopt the agent pull model step by step. Traditional applications use the Application Controller. Agent applications sync to workload clusters via agent pull model. Both appear in the same UI. Label selectors enforce isolation between them.
+
For more information, see link:https://argocd-agent.readthedocs.io/latest/user-guide/hybrid-architecture/[Hybrid Architecture] in the Argo CD Agent upstream documentation.
+
--
:FeatureName: The Argo CD Agent hybrid architecture feature
include::snippets/technology-preview.adoc[]
--
+

> link:https://issues.redhat.com/browse/GITOPS-9790[GITOPS-9790]

Reviewer 8: GITOPS-9022 is the story for hybrid architecture
```suggestion
link:https://issues.redhat.com/browse/GITOPS-9022[GITOPS-9022]
```

Author: Fixed as 9022.

+
Use the camelCase `logFormat` field for `applicationSet` and `notifications`. The lowercase `logformat` field is deprecated but still works for backward compatibility. The {gitops-title} Operator uses the lowercase `logformat` field if the camelCase `logFormat` field is not set. You can update your custom resource specifications to use the camelCase `logFormat` field. The deprecated lowercase `logformat` field might be removed in a future release.
+
link:https://issues.redhat.com/browse/GITOPS-9119[GITOPS-9119]

Deprecated dynamic shard scaling feature::
With this update, the dynamic shard scaling feature in the {gitops-title} Operator is deprecated. This includes the `.spec.controller.sharding.dynamicScalingEnabled` field in the Argo CD custom resource. For more information about sharding clusters, see the Additional resources section.
+
link:https://issues.redhat.com/browse/GITOPS-9478[GITOPS-9478]

> Removed kube-rbac-proxy dependency::

Reviewer 3: Can be changed to migrated

Author: Fixed.

// Module included in the following assembly:
//
// * release_notes/gitops-release-notes-1-21-0.adoc

:_mod-docs-content-type: REFERENCE

[id="gitops-release-notes-1-21-0_{context}"]
= Release notes for {gitops-title} 1.21.0

[role="_abstract"]

> {gitops-title} 1.21.0 is available on {OCP} 4.14, 4.16, 4.17, 4.18, 4.19, 4.20, 4.21, and 4.22.

Reviewer 9: Is 4.22 GA? If not lets mention it after the release

Author: Hey @varshab1210, the OCP 4.22 release is definitely out. Here's the link: https://docs.redhat.com/en/documentation/openshift_container_platform/4.22/html/release_notes/ocp-4-22-release-notes. We can include it for our product as well.

[id="errata-updates-1-21-0_{context}"]
== Errata updates

RHEA-XXXX:NNNN - {gitops-title} 1.21.0 enhancement update advisory::
Issued: 2026-NN-XX
+
The enhancements included in this release are documented in the following advisory:
+
* link:https://access.redhat.com/errata/RHEA-XXXX:NNNN[RHEA-XXXX:NNNN]

If you installed the {gitops-title} Operator in the default namespace, run the following command to view the container images in this release:

[source,terminal]
----
$ oc describe deployment gitops-operator-controller-manager -n openshift-gitops-operator
----

[id="new-features-1-21-0_{context}"]
== New features

Argo CD Image Updater (General Availability)::
With this update, Argo CD Image Updater is generally available. It updates container images automatically for Kubernetes workloads managed by Argo CD. It tracks image versions in applications. You can define these versions in `ImageUpdater` custom resources (CR). It updates images using parameter overrides through the Argo CD API or the Git write-back method. The Image Updater supports applications built with Kustomize or Helm.
+
link:https://issues.redhat.com/browse/GITOPS-9754[GITOPS-9754]

> GitOps Console Plugin resource pages (Technology Preview)::

Reviewer 9: @keithchong this is for OCP 4.19+ right?

Reviewer 2: yes

Reviewer 9: @Dhruv-Soni11 should we mention that in the docs?

+
This feature includes:
+
* List and details pages for Argo CD and Argo Rollouts resources
* Filters for health and sync status
* Visual views for Applications and ApplicationSets
* Rollout topology in the OpenShift Console Topology view
* YAML templates for creating resources
* Links between the OpenShift Console and the Argo CD UI
* Views for sync history, events, resources, generators, revisions, and project policies
+
These updates help you manage {gitops-shortname} resources from a single interface and reduce the need to switch between tools.
+
--
:FeatureName: The GitOps Console Plugin resource pages feature
include::snippets/technology-preview.adoc[]
--
+
link:https://issues.redhat.com/browse/GITOPS-9196[GITOPS-9196]

Argo CD CLI (General Availability)::
With this update, the Argo CD CLI is generally available. It provides a stable command-line interface for managing Argo CD applications and related resources. Use the CLI to perform common GitOps operations.
+
link:https://issues.redhat.com/browse/GITOPS-9379[GITOPS-9379]

> Multi-source Applications support in Argo CD UI::

Reviewer 9: @Dhruv-Soni11 please remove this

Author: Removed.

> Before this update, multi-source Applications (`spec.sources`) could only be created through the CLI, manifests, or manual YAML editing in the UI. The *New Application* panel supported only a single source. Users had to switch to YAML editing to combine multiple Git, Helm, or OCI sources. With this update, you can create applications with multiple sources directly from the New Application panel.

Reviewer 10: Hi Dhruv, we need to exclude this. This will be for ArgoCD 3.5 not 3.4
```suggestion

```

Author: Removed from content, fixed.

Automatic cleanup of legacy KAM resources::
With this update, {gitops-title} cleans up old KAM (Application Manager CLI) resources automatically. Before this update, you had to remove KAM components manually when upgrading from older versions. These resources often stayed in clusters after upgrades. This caused stale deployments, wasted resources, and security scan alerts. The {gitops-title} Operator removes these resources automatically.
+
link:https://issues.redhat.com/browse/GITOPS-9217[GITOPS-9217]

ServiceMonitor resources for Argo CD Agent::
With this update, ServiceMonitor resources are available for Argo CD Agent Principal and Agent components. This enables Prometheus to scrape metrics through Monitoring. Metrics now appear in Console dashboards.
+
link:https://issues.redhat.com/browse/GITOPS-9595[GITOPS-9595]

Per-component Prometheus metrics configuration::
With this update, you can configure the `interval` and `scrapeTimeout` fields for Argo CD component ServiceMonitors. This gives you control over metric collection intervals and timeouts for each Argo CD component.
+
link:https://issues.redhat.com/browse/GITOPS-9633[GITOPS-9633]

> Web-based CLI terminal for Argo CD::

Reviewer 9: @akhilnittala

Reviewer 9: @Dhruv-Soni11 this is a new feature but there is a known issue. Here is the detail: 

With this update, you can log in to a web-based CLI terminal in Argo CD by setting the `spec.webTerminalEnabled` field. This allows you to get a shell inside a running pod, similar to `kubectl exec`, and provides terminal access directly from the Argo CD web interface. 
When enabled, the terminal tab appears on the application details page for a pod, but command input does not work. The openshift-gitops-argocd-server service account does not have permission to create pods/exec resources, which the web terminal requires to attach to running pods.

To work around this issue, create a ClusterRole and ClusterRoleBinding that grants the Argo CD server service account pods/exec access. 

Example:

```
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: terminal-tmp
rules:
  - verbs:
      - create
    apiGroups:
      - ''
    resources:
      - pods/exec
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: terminal-tmp
subjects:
  - kind: ServiceAccount
    name: openshift-gitops-argocd-server
    namespace: openshift-gitops
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: terminal-tmp
```
Apply the manifest with oc apply -f <filename>. If your Argo CD instance is not the default deployment in openshift-gitops, update the service account name and namespace in the ClusterRoleBinding accordingly.

cc: @svghadi @anandf

Reviewer 5: @Dhruv-Soni11 - How will this be documented? A separate entry in release-note about the feature and then the workaround in known issues sections or everything in the known issue section?

Author: I think it will be confusing for the user to keep it as a new feature and a known issue. Hence, I want to keep this as a known issue only to avoid it being added it in multiple places.

Author: Fixed.

> With this update, you can log in to a web-based CLI terminal in Argo CD. This provides terminal access directly from the Argo CD web interface.

Reviewer 3: mention which field to set, you can login to web based cli terminal in Argo CD by setting spec.webTerminalEnabled field. WebTerminalEnabled allows you to get a shell inside a running pod just like you would with kubectl exec.

something like this. feel free to modify.

Author: Fixed.

Image Updater status subresource for monitoring and observability::
With this update, each `ImageUpdater` CR maintains a status subresource that reflects the overall state of the resource. This includes the number of matched applications, managed images, timestamps, conditions, and a list of image updates from the most recent update cycle. You can check the state of your `ImageUpdater` resources with CLI commands:
+
[source,terminal]
----
$ oc get imageupdater -n openshift-gitops
----
+
You can also view the complete status by examining the `ImageUpdater` CR in the cluster.
+
link:https://issues.redhat.com/browse/GITOPS-7316[GITOPS-7316]

[id="fixed-issues-1-21-0_{context}"]
== Fixed issues

Dex service account authentication security tokens::
Before this update, the Dex service account used non-expiring security tokens. This created security risks from long-lived credentials. With this update, the service account uses expiring tokens. This improves authentication security for the Dex server in {gitops-title}.
+
link:https://issues.redhat.com/browse/GITOPS-9426[GITOPS-9426]

Dex copyutil initContainer reconciliation during upgrades::

> Before this update, the Operator did not apply changes to the Dex `copyutil` initContainer during upgrades from older {gitops-title} versions. With this update, changes to the initContainer's `resources` and `volumeMounts` fields are applied correctly. This includes updates to the `dexconfig` volume mount and resource limits.

Reviewer 9: This is part of 1.20.4 already

Author: Removing this content as it is a part of the 1.20.4 release. Thanks, Varsha.

+
link:https://issues.redhat.com/browse/GITOPS-9549[GITOPS-9549]

> Service certificate annotation removal during upgrades::

Reviewer 9: This is part of 1.20.4 already

Author: Removing this content as it is a part of the 1.20.4 release. Thanks, Varsha.

+
link:https://issues.redhat.com/browse/GITOPS-9701[GITOPS-9701]

End-to-end TLS encryption for Redis traffic in Argo CD Agent and Principal::
With this update, the Argo CD Agent and Principal components support end-to-end TLS encryption for Redis traffic. The Redis proxy accepts TLS connections from Argo CD components like `argocd-server` and `argocd-repo-server`. It uses TLS when connecting to the local `argocd-redis` instance. Configure this feature using new CLI flags and environment variables. These include `--redis-tls-enabled`, proxy server certificate and key paths, or related Kubernetes secrets. You can configure the upstream CA using a file path, a Kubernetes secret, or an insecure mode for testing only. The Kubernetes and Helm installation manifests include optional Redis TLS settings. TLS is disabled by default. The `principal.redis.tls.enabled` and `agent.redis.tls.enabled` fields default to `false`. This ensures that existing deployments continue to work until you enable the feature and provide certificates.
+
link:https://issues.redhat.com/browse/GITOPS-8090[GITOPS-8090]

> 

Reviewer 11: ```suggestion

Redis credentials distribution using volume mounts::
With this update, username and password for the operator-managed Redis or Redis HA are injected using volume mounts. This is a security harnening making accidential credentials leaks less likely.

link:https://redhat.atlassian.net/browse/GITOPS-7368[GITOPS-7368]

```

@Dhruv-Soni11, please consider including this one. I have not nominated this initially, as it is completely invisible for the customers.

Author: Fixed.

+
link:https://issues.redhat.com/browse/GITOPS-9134[GITOPS-9134]

TLS encryption for Image Updater webhook server::
With this update, the Argo CD Image Updater webhook server runs with TLS enabled by default. Before this update, the webhook server accepted registry notifications over plain HTTP. It had no TLS encryption or configuration options. This allowed webhook payloads to transmit in cleartext. This included shared secrets used for payload validation. The webhook server now uses TLS 1.3 for both the minimum and maximum protocol versions by default. These TLS options are configurable. This ensures secure transmission of webhook notifications from container registries.
+
link:https://issues.redhat.com/browse/GITOPS-9133[GITOPS-9133]

> Declarative management for Git webhook secrets::

Reviewer 10: Thanks for adding this @Dhruv-Soni11
